tools/bridge/end2end-tests/tests/test_transfer.py

```diff
@@ -0,0 +1,427 @@
+                if time.time() - start_time > self.timeout:
+                    raise TimeoutException(
+                        f"Service {self.name} did not report to be up after {self.timeout} seconds"
+                    )
+                time.sleep(self.poll_interval)
+            else:
+                break
+
+    def terminate(self):
+        try:
+            self.process.terminate()
+            self.process.wait(timeout=10)
+        except subprocess.TimeoutExpired:
+            warnings.warn(f"{self.name} did not terminate in time and had to be killed")
+            self.process.kill()
+            self.process.wait()
+
+
+class Node(Service):
+    def __init__(self, path, *, name=None, port_shift=0):
+
+        super().__init__(
+            [
+                "parity",
+                "-d",
+                str(path),
+                "--config",
+                "dev-insecure",
+                "--ports-shift",
+                f"{port_shift}",
+            ],
+            name=name,
+        )
+        self.web3 = Web3(HTTPProvider(f"http://localhost:{8545+port_shift}"))
+
+    def is_up(self):
+        try:
+            version = self.web3.clientVersion
+            return bool(version)
+        except requests.exceptions.ConnectionError:
+            return False
+
+
+class Bridge(Service):
+    def __init__(
+        self,
+        *,
+        name=None,
+        validator_private_key,
+        token_contract_address,
+        foreign_bridge_contract_address,
+        home_bridge_contract_address,
+    ):
+        env = {
+            "HOME_RPC_URL": HOME_RPC_URL,
+            "FOREIGN_RPC_URL": FOREIGN_RPC_URL,
+            "FOREIGN_CHAIN_TOKEN_CONTRACT_ADDRESS": str(token_contract_address),
+            "FOREIGN_BRIDGE_CONTRACT_ADDRESS": str(foreign_bridge_contract_address),
+            "FOREIGN_CHAIN_EVENT_FETCH_START_BLOCK_NUMBER": "0",
+            "HOME_BRIDGE_CONTRACT_ADDRESS": str(home_bridge_contract_address),
+            "HOME_CHAIN_EVENT_FETCH_START_BLOCK_NUMBER": "0",
+            "VALIDATOR_PRIVATE_KEY": validator_private_key.to_hex(),
+            "HOME_RPC_TIMEOUT": "10",
+            "HOME_CHAIN_MAX_REORG_DEPTH": "0",
+            "HOME_CHAIN_EVENT_POLL_INTERVAL": "1",
+            "FOREIGN_RPC_TIMEOUT": "10",
+            "FOREIGN_CHAIN_MAX_REORG_DEPTH": "3",
+            "FOREIGN_CHAIN_EVENT_POLL_INTERVAL": "1",
+        }
+        env.update(os.environ)
+        super().__init__(["tlbc-bridge"], env=env, name=name)
+
+    def is_up(self):
+        # TODO write uptest for the bridge
+        return True
+
+
+def mine_blocks(web3, number_of_blocks):
+    """Mines `number_of_blocks` blocks. Works by sending out transactions, assumes the parity InstaSeal engine"""
+
+    block_height = web3.eth.blockNumber
+    while web3.eth.blockNumber < block_height + number_of_blocks:
+        web3.eth.sendTransaction({"from": PARITY_DEV_ACCOUNT})
+
+    assert web3.eth.blockNumber == block_height + number_of_blocks
```

Comment:
I guess if there are two transactions sent simultaneously, this assert could fail. So maybe just use >=?

Reply:
This check is exactly to check this, as this function should mine exactly the number of blocks. This function only works if there are no other transaction at the same time. 
We could change it to `mine_min_blocks` or so and remove the check.

Reply:
Changed it to `mine_min_blocks`